### PR TITLE
Factory Pattern for All the API Calls

### PR DESCRIPTION
**Pull request by NJaryal**
Initial Commit - Single Api Call for all the Methods

### CODE REVIEW

File: src/js/component/Api.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+       }
+     return fetch(this.path + url, param)
+   }
+     
+   get(url) {
+     return this.request(url, 'GET')
+   }  
+     
+   post(url, body) {
+    // code here
```
Comment (byslavik) [inline src/js/component/Api.js]: put some code here :)

File: src/js/component/Api.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+   }   
+     
+   put(url, body) {
+     // code here
```
Comment (byslavik) [inline src/js/component/Api.js]: and here :)

File: src/js/component/App.js [hunk 3]
```diff
@@ -22,6 +23,7 @@ class App {
     this.ul = document.querySelector(settings.ulSelector)
     this.channelUILists = document.querySelector(settings.channelUIListsSelector)    
     this.channels = {}
+    this.api = new Api("")
```
Comment (byslavik) [inline src/js/component/App.js]: I will suggest to use one instance of api for all appication.
for example
class Api {
// some code here
}

export const api = new Api(BASE_URL + API_KEY) // by writing that string all pieces of your app will use single instance which making calls to single url

Then you can just import it as api and use as api.get('some addditinal url') api.put() api.post()

File: src/js/component/App.js [hunk 4]
```diff
@@ -50,13 +52,13 @@ class App {
   async get(source) {
     this.spinner.enableSpinner()    
       try {
-        const response = await fetch(`${BASE_URL}${source}&apiKey=${API_KEY}`);
+        const response = await this.api.get(`${BASE_URL}${source}&apiKey=${API_KEY}`);
```
Comment (byslavik) [inline src/js/component/App.js]: as per my comment above it will be just like api.get(source)

File: src/js/component/App.js (hunk 3 above)
Comment (NJaryal) [reply src/js/component/App.js]: I applied same way last evening - By taking our Main App reference file. But I had been getting some errors. I agree and let me try again.

File: src/js/component/Api.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+class Api {
+    constructor(path) {
+      this.path = path
+    }
+   
+   request(url, method, body) {
```
Comment (byslavik) [inline src/js/component/Api.js]: Where is fetch from here?

File: src/js/component/Api.js [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+       const param = {
+           method,
+           body: JSON.stringify(body)
+       }
+       return fetch(this.path + url, param)
```
Comment (byslavik) [inline src/js/component/Api.js]: fetch(this.path + url, param).then(res => res.json())

File: src/js/component/Api.js (hunk 2 above)
Comment (NJaryal) [reply src/js/component/Api.js]: Updated

File: src/js/component/Api.js (hunk 1 above)
Comment (NJaryal) [reply src/js/component/Api.js]: Updated

File: src/js/component/App.js (hunk 3 above)
Comment (NJaryal) [reply src/js/component/App.js]: It works and Applied as suggested

File: src/js/component/Api.js (hunk 5 above)
Comment (NJaryal) [reply src/js/component/Api.js]: Excuse me - It was just there 
Hidden

File: src/js/component/Api.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import {Errors} from './Errors';
+import {view} from '../../Views/View'
+import { settings} from './Variable';
+class Api {
+    constructor(path, settings) {
+      this.path = path 
+      this.settings = settings
+      this.errorsModal = document.querySelector(settings.errorsModal)
```
Comment (byslavik) [inline src/js/component/Api.js]: better to define that stuff in view and use like view.setError(err.message)

File: src/js/component/App.js [hunk 8]
```diff
@@ -20,7 +23,8 @@ class App {
     this.main = document.querySelector(settings.gridSection)
     this.nav = document.querySelector(settings.navSection)
     this.ul = document.querySelector(settings.ulSelector)
-    this.channelUILists = document.querySelector(settings.channelUIListsSelector)    
+    this.channelUILists = document.querySelector(settings.channelUIListsSelector)
```
Comment (byslavik) [inline src/js/component/App.js:19]: your view should take care of that

Comment (byslavik) [reply src/js/component/App.js:19]: https://www.sitepoint.com/mvc-design-pattern-javascript/ here is an example of how it usually used

File: src/js/component/Api.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+     
+   post(url, body) {
+    return this.request(url, 'POST', body)
+   }   
+     
+   put(url, body) {
+    return this.request(url, 'PUT', body)
+   }
+}
+export const api = new Api('', settings);
```
Comment (byslavik) [inline src/js/component/Api.js]: try to put base url as the first parameter to avoid putting of that in multiple places

File: src/js/component/Proxy.js [hunk 10]
```diff
@@ -1,3 +1,13 @@
+import {api} from './Api'
+const proxyHandler = {
```
Comment (byslavik) [inline src/js/component/Proxy.js:3]: what is the purpose of your proxy? we need to log reuests and responses :)

File: src/js/component/App.js [hunk 11]
```diff
@@ -34,33 +38,19 @@ class App {
...
-          </div>
-        </div>`   
+      this.nav.innerHTML += view.navHTML(item)
+      this.main.innerHTML += view.mainHTML(item)
     }); 
   }  
 
-  async get(source) {
+  get(source) {
     this.spinner.enableSpinner()
```
Comment (byslavik) [inline src/js/component/App.js]: spinner also should be delegated to the view

File: src/js/component/Proxy.js [hunk 12]
```diff
@@ -1,3 +1,13 @@
...
+  get: function(target, name){
+        return target[name]
+    }
+}
+const proxiedApi = new Proxy(api, proxyHandler)
+proxiedApi.request()
+proxiedApi.get()
```
Comment (byslavik) [inline src/js/component/Proxy.js]: proxiedApi.request()
proxiedApi.get()

Why do you need it here? It was just an example of usage. With current implementation no one of your requests will not be logged via such proxy. You need to export proxied api and use it as api in all places

File: src/js/component/App.js (hunk 11 above)
Comment (NJaryal) [reply src/js/component/App.js]: Delegated as a separate Class inside the Views

File: src/js/component/Api.js [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    }
+   
+   async request(url, method, body) {
+      const param = {
+          method,
+          body: JSON.stringify(body)
+      }
+      try {     
+        return await fetch(this.path + url, param).then(res => res.json())
+        throw Errors.getInstance()
```
Comment (byslavik) [inline src/js/component/Api.js]: unreachable piece of code

File: src/js/component/Errors.js [hunk 14]
```diff
@@ -0,0 +1,16 @@
+import {view} from '../../Views/View';
+export class Errors {
```
Comment (byslavik) [inline src/js/component/Errors.js:2]: Currently this is not a singleton. This is just a class

File: src/js/component/Headlines.js [hunk 15]
```diff
@@ -1,15 +1,18 @@
+import {view} from '../../Views/View'
+import{headlinesNewsData} from './LazyLoading'
```
Comment (byslavik) [inline src/js/component/Headlines.js]: please fix spacing

File: src/js/component/Headlines.js [hunk 16]
```diff
@@ -1,15 +1,18 @@
...
+      return await import(/* webpackPreload: true */ "./LazyLoading.js")
+      .then(({ articles }) => {
           articles.map(item => {
-            this.modalHeadlines.innerHTML += `<a href="#" class="list-group-item"><strong>${item.title} </strong></li>`;
+            this.modalHeadlines.innerHTML += view.headLinesHTML(item);
           });
         });
     }
 }
+const newsHeadings = new Headlines(settings)
```
Comment (byslavik) [inline src/js/component/Headlines.js]: Currently your headlines are loaded after opening of the site. Could you please fix it to load only when modal is shown?

File: src/js/component/LazyLoading.js [hunk 17]
```diff
@@ -1,10 +1,15 @@
...
+import {proxiedApi} from './Proxy';
+const headlinesNewsData = () => {
+    /* let api = new Api('') */
     try {
-        const resp = await fetch(headLines_KEY);
-        return await resp.json();  
-    } catch (error) {
-        console.log(error);
+        return proxiedApi.get(headLines_KEY);
+        throw Errors.getInstance()
```
Comment (byslavik) [inline src/js/component/LazyLoading.js]: unreachable piece of code

File: src/js/component/Api.js [hunk 18]
```diff
@@ -0,0 +1,34 @@
...
+          method,
+          body: JSON.stringify(body)
+      }
+      try {
+        return await fetch(this.BASE_URL + url , param).then(res => {
+          console.log(`Status: ${res.status} Url:${res.url}`)
+          res.json()})
+        throw Errors.getInstance()
+      } catch (e) {
+          import("./Errors.js");
```
Comment (byslavik) [inline src/js/component/Api.js]: That will just import your module, but not do anything with it

File: src/Views/Spinner.js [hunk 19]
```diff
@@ -1,6 +1,7 @@
+import { settings} from '../js/component/Variable';
```
Comment (byslavik) [inline src/Views/Spinner.js]: 'component/Variable' better to place it as a separate file not in scope of components. File with constants should be named as constants.js 
All files which are related to components/classes should be named with the first letter in upper case, all other files in lover case:
examples - App, Spinner, NewsChannel but helpers, constants, selectors....

File: src/js/component/App.js [hunk 20]
```diff
@@ -30,52 +27,30 @@ class App {
...
-        `<div class="thumbnail">
-          <div class="caption">
-          <h3>${item.label}</h3>
-          <p><a href="#" class="btn btn-primary" data-load-channel="${item.inst}" role="button">${item.source}</a></p>
-          </div>
-        </div>`   
-    }); 
-  }  
+  get(source) {
+    spinner.enableSpinner()
```
Comment (byslavik) [inline src/js/component/App.js]: can we do smth like view.spinner.enableSpinner()?

File: src/js/component/Api.js [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+    }
+       
+   async request(url, method, body) {
+      const param = {
+          method,
+          body: JSON.stringify(body)
+      }
+      try {
+        return await fetch(this.BASE_URL + url , param).then(res => {
+          console.log(`Status: ${res.status} Url:${res.url}`)
```
Comment (byslavik) [inline src/js/component/Api.js]: Lets move it to proxy

Comment (NJaryal) [reply src/js/component/Api.js]: I understood little late
Console.log shifted to proxy

**Review by NJaryal — COMMENTED** [review]
Updated + resolved all the valuable comments

File: src/js/component/App.js [hunk 22]
```diff
@@ -30,52 +26,30 @@ class App {
...
-          <div class="caption">
-          <h3>${item.label}</h3>
-          <p><a href="#" class="btn btn-primary" data-load-channel="${item.inst}" role="button">${item.source}</a></p>
-          </div>
-        </div>`   
-    }); 
-  }  
+  get(source) {
+    view.enableSpinner()   
+    return proxiedApi.get(`${BASE_URL}${source}&apiKey=${API_KEY}`)
```
Comment (byslavik) [inline src/js/component/App.js:33]: do we need base url here if we have it in Api?

**Comment by NJaryal** [comment]
Merging the request after doing the changes

File: src/Views/View.js [hunk 23]
```diff
@@ -0,0 +1,75 @@
+import {sources,settings} from "../../src/js/constant/constant";
+import "../js/component/BootstrapMenu";
+class Views {
+  constructor(sources, settings) {
+    this.sources = sources;
+    this.settings = settings;
+    this.errorsModal = document.querySelector(settings.errorsModal)
+    this.main = document.querySelector(settings.gridSection)
+    this.nav = document.querySelector(settings.navSection)
+    this.errorsModal = document.querySelector(settings.errorsModal)
```
Comment (byslavik) [inline src/Views/View.js:10]: looks like redundand assignement
